File: core/blas/strmv.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "strmv.h"
#include "blas_common.h"
#include "blasplan/include/blasplan/BlasStrmvPlan.h"

using namespace AsdSip;
using namespace Mki;

namespace AsdSip {
struct StrmvTensorParam {
    const aclTensor *A;
    const aclTensor *x;
};

AspbStatus StrmvDtypeCheck(struct StrmvTensorParam parm)
{
    auto ret = ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE;
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.A, aclDataType::ACL_FLOAT, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.x, aclDataType::ACL_FLOAT, ret);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus StrmvShapeCheck(struct StrmvTensorParam parm, const int64_t n)
{
    ASDSIP_ECHECK(
        n > 0 && n <= UINT32_MAX, "blas asdBlasStrmv get n <= 0 or n > 2^32.", ErrorType::ACL_ERROR_INVALID_PARAM);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.A, n * n, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.x, n, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus asdBlasStrmv(asdBlasHandle handle, asdBlasFillMode_t uplo, asdBlasOperation_t trans, asdBlasDiagType_t diag,
    const int64_t n, aclTensor *A, const int64_t lda, aclTensor *x, const int64_t incx)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(BlasPlanCache::doesPlanExist(handle),
        "blas Strmv get cached plan failed.",
        ErrorType::ACL_ERROR_INTERNAL_ERROR);

    struct StrmvTensorParam parm {
        A, x
    };
    ASDSIP_CHECK(StrmvDtypeCheck(parm) == ErrorType::ACL_SUCCESS,
        "blas asdBlasStrmv dtype check failed.",
        return ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    auto ret = StrmvShapeCheck(parm, n);
    ASDSIP_CHECK(ret == ErrorType::ACL_SUCCESS, "blas asdBlasStrmv shape check failed.", return ret);

    if (incx <= 0) {
        ASDSIP_LOG(INFO) << "blas asdBlasStrmv get incx <= 0.";
    }

    try {
        AsdSip::BlasStrmvPlan &plan = dynamic_cast<AsdSip::BlasStrmvPlan &>(BlasPlanCache::getPlan(handle));
        ASDSIP_ECHECK(plan.IsInitialized(), "Strmv plan init Error!.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

        OpDesc opDesc;
        opDesc.opName = "StrmvOperation";
        AsdSip::OpParam::Strmv param = {0, 0, 0, 0, 0};

        if (trans == asdBlasOperation_t::ASDBLAS_OP_N) {
            param.trans = 0;
        } else {
            param.trans = 1;
        }

        if (uplo == asdBlasFillMode_t::ASDBLAS_FILL_MODE_UPPER) {
            param.uplo = 1;
        } else {
            param.uplo = 0;
        }

        if (diag == asdBlasDiagType_t::ASDBLAS_DIAG_UNIT) {
            param.diag = 1;
        } else {
            param.diag = 0;
        }
        param.incx = 1;
        param.lda = n;
        opDesc.specificParam = param;
        ASDSIP_LOG(DEBUG) << "OpDesc: " << opDesc.opName << "; OpDesc info: " << param.ToString();

        SVector<aclTensor *> inTensors{A, x, plan.GetmMaskAclTensor()};
        SVector<aclTensor *> outTensors{x};

        Status status = RunAsdOpsV2(plan.GetStream(), opDesc, inTensors, outTensors, plan.GetWorkspace());
        ASDSIP_ECHECK(status.Ok(), status.Message(), ErrorType::ACL_ERROR_INTERNAL_ERROR);

        ASDSIP_LOG(INFO) << "Execute asdBlasStrmv success.";
        return ErrorType::ACL_SUCCESS;
    } catch (std::bad_cast &e) {
        // Handle the op strmv exception
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Strmv Error: " << e.what();
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
}

AspbStatus asdBlasMakeStrmvPlan(asdBlasHandle handle, asdBlasFillMode_t uplo, asdBlasOperation_t trans, int64_t n)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(handle != nullptr, "blas MakeStrmvPlan Fail.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

    AsdSip::BlasStrmvPlan *plan = nullptr;
    try {
        plan = new AsdSip::BlasStrmvPlan(uplo, trans, n);
        BlasPlanCache::MakePlan(handle, plan);
    } catch (const std::exception &e) {
        if (plan != nullptr) {
            delete plan;
        }
        delete static_cast<int *>(handle);
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Make Strmv Plan failed: " << e.what();
        throw std::runtime_error("Make Strmv Plan failed.");
    }
    if (plan->CreateTensor() != ErrorType::ACL_SUCCESS) {
        plan->MarkFailed();
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Fail to create blas asdBlasMakeStrmvPlan mask tensor.";
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
    plan->MarkInitialized();
    return ErrorType::ACL_SUCCESS;
}
}  // namespace AsdSip

File: sip_pta/test/blas/strmv.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Copyright (c) 2026 Huawei Technologies Co., Ltd.
This program is free software, you can redistribute it and/or modify it under the terms and conditions of
CANN Open Software License Agreement Version 2.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
"""

import logging
import os
import sys
from dataclasses import dataclass

import torch
import torch_npu
import torch_sip

# 配置全局日志 (G.LOG.02)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[logging.StreamHandler(sys.stdout)]
)
logger = logging.getLogger("torch_sip_strmv_test")


@dataclass(frozen=True)
class TrmvConfig:
    """参数对象：封装 STRMV 的所有维度、模式及名称配置"""
    n: int
    uplo: str = "L"
    trans: str = "N"
    diag: str = "N"
    name: str = "Case"


class StrmvTester:
    """单精度三角矩阵-向量乘法 (STRMV) 算子测试类"""

    def __init__(self, device: str = "npu:0"):
        self.device = device
        blocking_env = os.getenv("ASCEND_LAUNCH_BLOCKING", "0")
        mode_str = "异步 (Async)" if blocking_env == "0" else "同步 (Blocking)"
        logger.info("当前 NPU 运行模式: %s (ASCEND_LAUNCH_BLOCKING=%s)", mode_str, blocking_env)

    @staticmethod
    def to_column_major(tensor: torch.Tensor) -> torch.Tensor:
        """确保矩阵满足列优先布局需求 (静态方法)"""
        return tensor.t().contiguous().t()

    def run_case(self, cfg: TrmvConfig) -> bool:
        """
        版本 1: A 由行存储转为列存储。
        G.ERR.01: 最小化 try 块。
        """
        # 1. 准备数据
        mat_a = torch.randn((cfg.n, cfg.n), dtype=torch.float32, device=self.device)
        vec_x = torch.randn((cfg.n,), dtype=torch.float32, device=self.device)
        mat_a = self.to_column_major(mat_a)

        # 2. CPU 参考计算逻辑 (逻辑拆分)
        a_ref = mat_a.cpu().clone()
        if cfg.uplo.upper() == "L":
            a_ref = torch.tril(a_ref)
        else:
            a_ref = torch.triu(a_ref)

        if cfg.diag.upper() == "U":
            a_ref.fill_diagonal_(1.0)

        if cfg.trans.upper() == "T":
            a_ref = a_ref.t()

        ref_res = torch.mv(a_ref, vec_x.cpu())

        # 3. 核心调用 (G.ERR.01: 仅包裹原地修改的算子执行)
        try:
            torch_sip.asd_blas_strmv(mat_a, vec_x, cfg.uplo, cfg.trans, cfg.diag)
        except Exception as exc:
            logger.error("[%s] V1 算子崩溃: %s", cfg.name, exc)
            return False

        # 4. 精度比对
        is_close = torch.allclose(vec_x.cpu(), ref_res, rtol=1e-3, atol=1e-3)
        status = "PASS" if is_close else "FAIL"
        logger.info("[%s] V1 | N=%d | %s, %s, %s | %s",
                    cfg.name, cfg.n, cfg.uplo, cfg.trans, cfg.diag, status)
        return is_close

    def run_case_v2(self, cfg: TrmvConfig) -> bool:
        """
        版本 2: A 不转为列存储, 使用 A^T 传入算子。
        G.ERR.01: 最小化 try_块。
        """
        # 1. 准备数据
        mat_a_raw = torch.randn((cfg.n, cfg.n), dtype=torch.float32, device=self.device)
        vec_x = torch.randn((cfg.n,), dtype=torch.float32, device=self.device)

        # 2. CPU 参考计算逻辑
        a_ref = mat_a_raw.cpu().clone()
        mat_a_npu = mat_a_raw.t()  # 模拟转置输入

        if cfg.uplo.upper() == "L":
            a_ref = torch.triu(a_ref)  # 转置后 L 变 U
        else:
            a_ref = torch.tril(a_ref)  # 转置后 U 变 L

        if cfg.diag.upper() == "U":
            a_ref.fill_diagonal_(1.0)

        if cfg.trans.upper() != "T":
            a_ref = a_ref.t()

        ref_res = torch.mv(a_ref, vec_x.cpu())

        # 3. 核心调用
        try:
            torch_sip.asd_blas_strmv(mat_a_npu, vec_x, cfg.uplo, cfg.trans, cfg.diag)
        except Exception as exc:
            logger.error("[%s] V2 算子崩溃: %s", cfg.name, exc)
            return False

        # 4. 精度比对
        is_close = torch.allclose(vec_x.cpu(), ref_res, rtol=1e-3, atol=1e-3)
        status = "PASS" if is_close else "FAIL"
        logger.info("[%s] V2 | N=%d | %s, %s, %s | %s",
                    cfg.name, cfg.n, cfg.uplo, cfg.trans, cfg.diag, status)
        return is_close


def main():
    """测试主套件入口"""
    tester = StrmvTester()
    logger.info("开始 STRMV 专项测试 (参数对象化重构版本) ...\n")

    # 定义测试套件
    test_suites = [
        TrmvConfig(4, "U", "N", "N", "Upper_Normal_NonUnit"),
        TrmvConfig(8, "L", "N", "N", "Lower_Normal_NonUnit"),
        TrmvConfig(16, "U", "T", "N", "Upper_Trans_NonUnit"),
        TrmvConfig(32, "L", "T", "U", "Lower_Trans_Unit"),
        TrmvConfig(64, "U", "N", "U", "Upper_Normal_Unit"),
    ]

    all_passed = True

    # 批量执行 V1 测试
    logger.info("--- 执行 V1 (Column-Major) 测试 ---")
    for config in test_suites:
        if not tester.run_case(config):
            all_passed = False

    # 批量执行 V2 测试
    logger.info("\n--- 执行 V2 (Row-Major/Transposed) 测试 ---")
    for config in test_suites:
        if not tester.run_case_v2(config):
            all_passed = False

    logger.info("-" * 60)
    if all_passed:
        logger.info("测试结论: ✅ 全部通过")
    else:
        logger.error("测试结论: ❌ 存在失败项")

    return 0 if all_passed else 1


if __name__ == "__main__":
    sys.exit(main())